### PR TITLE
Add metadata in mc copy

### PR DESCRIPTION
**Pull request by sinhaashish**
mc copy provisioned to pass metadata as string.

Example : mc cp --metadata key1=value1,key=value2 play/mybucket/burningman2011/ ~/latest/

Fixes: [2639](https://github.com/minio/mc/issues/2636)

### CODE REVIEW

**Comment by oliverkane** [comment]
Subscribing to this as this looks useful for a personal project of mine.

File: functional-tests.sh [hunk 1]
```diff
@@ -334,6 +334,20 @@ function test_put_object_with_storage_class_error()
...
 }
 
+## Test mc cp command with metadata set to a json string
+function test_put_object_with_metadata()
+{
+    show "${FUNCNAME[0]}"
+
+    start_time=$(get_time)
+    object_name="mc-test-object-$RANDOM"
+    assert_failure "$start_time" "${FUNCNAME[0]}" mc_cmd cp --attr '{"key1":"value1","key1":"value1"}' "${FILE_1_MB}" "${SERVER_ALIAS}/${BUCKET_NAME}/${object_name}"
```
Comment (harshavardhana) [inline functional-tests.sh]: This functional test is incomplete, we should also add `mc stat` to validate if the attribute is set.

File: cmd/cp-main.go [hunk 2]
```diff
@@ -438,6 +464,29 @@ loop:
 	return retErr
 }
 
+// parse the passed MetaData and populate the map
+func parseMetaData(metadataString string) (map[string]string, error) {
```
Comment (harshavardhana) [inline cmd/cp-main.go]: Can be simplified

```go
func parseMetaData(metadataString string) (map[string]string, error) {
	metaDataMap := make(map[string]string)
	for _, metaData := range strings.Split(metadataString, ",") {
		metaDataEntry := strings.Split(metaData, "=")
		if len(metaDataEntry) == 2 {
			metaDataMap[metaDataEntry[0]] = metaDataEntry[1]
			continue
		} else {
			return nil, errors.New("Unable to parse metadata")
		}
	}
}
```

File: cmd/cp-main.go [hunk 3]
```diff
@@ -359,14 +368,31 @@ func doCopySession(session *sessionV8, encKeyDB map[string][]prefixSSEPair) erro
 				// Save totalSize.
 				cpURLs.TotalSize = session.Header.TotalBytes
 
-				// Check and handle storage class if passed in command line args
-				if _, ok := session.Header.CommandStringFlags["storage-class"]; ok {
+				t, storageclassPresent := session.Header.CommandStringFlags["storage-class"]
+				fmt.Print(" ashishshshshshshshshsshshshshshshsh", t)
```
Comment (harshavardhana) [inline cmd/cp-main.go]: Remove debug statements @sinhaashish before submitting a PR as practice.

File: cmd/cp-main.go [hunk 4]
```diff
@@ -359,14 +368,31 @@ func doCopySession(session *sessionV8, encKeyDB map[string][]prefixSSEPair) erro
 				// Save totalSize.
 				cpURLs.TotalSize = session.Header.TotalBytes
 
-				// Check and handle storage class if passed in command line args
-				if _, ok := session.Header.CommandStringFlags["storage-class"]; ok {
+				t, storageclassPresent := session.Header.CommandStringFlags["storage-class"]
```
Comment (harshavardhana) [inline cmd/cp-main.go]: s/storageClassPresent/scSet/g is better to be concise..

File: cmd/cp-main.go [hunk 5]
```diff
@@ -359,14 +368,31 @@ func doCopySession(session *sessionV8, encKeyDB map[string][]prefixSSEPair) erro
...
+				fmt.Print(" ashishshshshshshshshsshshshshshshsh", t)
+				fmt.Print("\n \n  ashishshshshshshshshsshshshshshshsh \n \n ", t)
+
+				metaDataString, metadataPresent := session.Header.CommandStringFlags["metadata"]
```
Comment (harshavardhana) [inline cmd/cp-main.go]: s/metadataPresent/metaSet/g is better to be concise..

File: cmd/cp-main.go [hunk 6]
```diff
@@ -459,6 +508,7 @@ func mainCopy(ctx *cli.Context) error {
 		sseKeys = key
 	}
 	sse := ctx.String("encrypt")
+	userMetadata := ctx.String("attr")
```
Comment (harshavardhana) [inline cmd/cp-main.go]: We should always parse the input here, i.e parseMetadata should be used here.. look at checkCopySyntax or getEncKeys for adding validations..

File: cmd/cp-main.go [hunk 7]
```diff
@@ -57,6 +58,10 @@ var (
 			Name:  "encrypt",
 			Usage: "encrypt/decrypt objects (using server-side encryption with server managed keys)",
 		},
+		cli.StringFlag{
+			Name:  "attr",
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: It was discussed and thus `attr` is used. You can find in the comments of [issue](https://github.com/minio/mc/issues/2636)

File: cmd/cp-main.go [hunk 8]
```diff
@@ -438,12 +454,34 @@ loop:
...
 }
 
+// validate the passed metadataString and populate the map
+func getMetaDataEntry(metadataString string) (map[string]string, *probe.Error) {
+	metaDataMap := make(map[string]string)
+	for _, metaData := range strings.Split(metadataString, ",") {
+		metaDataEntry := strings.Split(metaData, "=")
+		if len(metaDataEntry) == 2 {
+			metaDataMap[metaDataEntry[0]] = metaDataEntry[1]
+			continue
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: removed.

File: cmd/cp-main.go [hunk 9]
```diff
@@ -438,12 +454,34 @@ loop:
...
+		} else {
+			return nil, probe.NewError(errors.New("specified metadata should be of form key1=value1,key2=value2,... and so on"))
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: Added

File: cmd/cp-main_test.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	//"errors"
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main_test.go]: Removed

File: cmd/cp-main.go [hunk 11]
```diff
@@ -438,12 +454,34 @@ loop:
...
 func mainCopy(ctx *cli.Context) error {
 	// Parse encryption keys per command.
 	encKeyDB, err := getEncKeys(ctx)
 	fatalIf(err, "Unable to parse encryption keys.")
 
+	// Parse metadata.
+	metaMap := make(map[string]string)
+	if ctx.String("attr") != "" {
+		metaMap, err = getMetaDataEntry(ctx.String("attr"))
+		fatalIf(err, "Unable to parse metadata.")
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: modified

File: cmd/cp-main.go [hunk 12]
```diff
@@ -359,14 +368,21 @@ func doCopySession(session *sessionV8, encKeyDB map[string][]prefixSSEPair) erro
 				// Save totalSize.
 				cpURLs.TotalSize = session.Header.TotalBytes
 
-				// Check and handle storage class if passed in command line args
-				if _, ok := session.Header.CommandStringFlags["storage-class"]; ok {
+				_, scSet := session.Header.CommandStringFlags["storage-class"]
+				metaMap, metaSet := session.Header.MetaData["metadata"]
+
+				// Check and handle metadata or storage class  if passed in command line args
+				if scSet || metaSet {
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: Done.

File: cmd/cp-main.go [hunk 13]
```diff
@@ -107,7 +112,11 @@ EXAMPLES:
 
    9. Copy a folder with encrypted objects recursively from Amazon S3 to Minio cloud storage.
       $ {{.HelpName}} --recursive --encrypt-key "s3/documents/=32byteslongsecretkeymustbegiven1,myminio/documents/=32byteslongsecretkeymustbegiven2" s3/documents/ myminio/documents/
-`,
+
+  10. Copy files and set the metadata  provided as json string.
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: updated

[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: updated

File: cmd/cp-main.go [hunk 14]
```diff
@@ -57,6 +58,10 @@ var (
 			Name:  "encrypt",
 			Usage: "encrypt/decrypt objects (using server-side encryption with server managed keys)",
 		},
+		cli.StringFlag{
+			Name:  "attr",
+			Usage: "add custom metadata string",
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/cp-main.go]: updated

File: cmd/cp-main.go [hunk 15]
```diff
@@ -57,9 +58,16 @@ var (
...
 		},
+		cli.StringFlag{
+			Name:  "attr",
+			Usage: "add custom metadata for the object",
+		},
 	}
 )
 
+// ErrInvalidMetadata reflects invalid metadata format
+var ErrInvalidMetadata = probe.NewError(errors.New("specified metadata should be of form key1=value1,key2=value2,... and so on"))
```
Comment (harshavardhana) [inline cmd/cp-main.go]: This is not a correct style of error when using probe package for example look at errDummy() implementation, you should use that style. 

Using probe in this manner adds the line number of this definition meaning mc --debug will show error is originating here which is not we need to stop at the function return.

Comment (harshavardhana) [reply cmd/cp-main.go]: Also the error to point to the exact position to indicate the error, basically use the input Arg and show in which position you see this problem. If that's not possible we can simply show the entire arg and indicate it doesn't follow expected style.

Comment (krisis) [reply cmd/cp-main.go]: OK, so can we have, `var ErrInvalidMetadata = errors.New("the appropriate error message")` and call `probe.NewError(ErrInvalidMetadata)` at the respective call sites?

File: cmd/cp-main.go [hunk 16]
```diff
@@ -367,6 +379,20 @@ func doCopySession(session *sessionV8, encKeyDB map[string][]prefixSSEPair) erro
...
 				}
 
+				metaMap, metaSet := session.Header.MetaData["metadata"]
+
+				// Check and handle metadata if passed in command line args
+				if metaSet {
+					if cpURLs.TargetContent.Metadata == nil {
+						cpURLs.TargetContent.Metadata = make(map[string]string)
+					}
+					if metaSet && len(metaMap) != 0 {
```
Comment (krisis) [inline cmd/cp-main.go]: we don't need to check `metaSet` again.

File: cmd/cp-main_test.go [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+	metaDataCases := []struct {
+		input  string
+		output map[string]string
+		err    error
+		status bool
+	}{
+		// success scenerio
+		{"key1=value1,key2=value2", map[string]string{"key1": "value1", "key2": "value2"}, nil, true},
+		// using different delimitter, other than ',' between multiple meta data
+		{"key1=value1;key2=value2", nil, errors.New("specified metadata should be of form key1=value1,key2=value2,... and so on"), false},
```
Comment (krisis) [inline cmd/cp-main_test.go]: we must be able to reuse `ErrInvalidMetadata` in the tests instead of `errors.New(..)`.

File: cmd/cp-main.go (hunk 15 above)
Comment (harshavardhana) [reply cmd/cp-main.go]: yes that is possible @krisis

File: cmd/cp-main_test.go (hunk 17 above)
Comment (sinhaashish) [reply cmd/cp-main_test.go]: done.

File: cmd/cp-main.go (hunk 16 above)
Comment (sinhaashish) [reply cmd/cp-main.go]: removed `metsSet`

**Comment by sinhaashish** [comment]
@Praveenrajmani  @krisis  PTAL

File: cmd/cp-main.go [hunk 18]
```diff
@@ -107,7 +115,11 @@ EXAMPLES:
 
    9. Copy a folder with encrypted objects recursively from Amazon S3 to Minio cloud storage.
       $ {{.HelpName}} --recursive --encrypt-key "s3/documents/=32byteslongsecretkeymustbegiven1,myminio/documents/=32byteslongsecretkeymustbegiven2" s3/documents/ myminio/documents/
-`,
+
+  10. Copy files and set the metadata provided as a string.
```
Comment (krisis) [inline cmd/cp-main.go]: The description says, "Copy files ...", but the example command copies from an object store to another, also recursively. We should provide two examples,
1. for copying a file to an object store
2. for copying an object from an object store to another.

On a related note, what happens when the source object already has some metadata set on it matching the supplied key-value pairs via `attr` flag?

Comment (sinhaashish) [reply cmd/cp-main.go]: Added the examples for the above scenario.
>On a related note, what happens when the source object already has some metadata set on it matching the supplied key-value pairs via attr flag?

The value is overridden with the new value passed as the parameter in `mc --attr`

**Comment by sinhaashish** [comment]
Added the code to append the metadata in case of server side copy with `--attr` flag
```  
1.  mc cp  --attr key1=val1 <<local_folder>> myminio/test/one 
2. mc cp --attr key2=val2 myminio/test/one myminio/test/two 
```
Object `two` has both `key1=val1`  and `key2=val2` as its metadata

@krisis @Praveenrajmani @harshavardhana  PTAL

**Pull request merged**